README.md
# hex-bin


Clojure/CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2020-03-02
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2020-03-02
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/hex-bin/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/hex-bin/compare/0.1.0...0.1.1


Clojure/test/porkostomus/hex_bin_test.clj
(ns porkostomus.hex-bin-test
  (:require [clojure.test :refer :all]
            [porkostomus.hex-bin :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


Python/binHex.py
from random import randint

def squares(rows):
    sqList = []
    n = 0
    for l in range(rows):
        for m in range(rows):
            sqList.append((n, str(l) + str(m)))
            n +=1
    return sqList
  
#print(squares(5))

def shuffle(l):
    count = 0
    randomList = []
    for i in range(0, (len(l))):
        number = randint(0, (len(l)-1))
        randomList.append(l.pop(number))
        count +=1
    return randomList

randomBoxStringTupleList = shuffle(squares(5))

def binHexNums():
    hexList = []
    binList = []
    binHexList = []
    aList = []
    hexString = '0x'
    secondDigit = ['0', 'a', 'f']
    num = 0
    count = 0

    for j in range(len(secondDigit)):
        for i in range(16):
            firstDigit = str(hex(i))
            firstDigit = firstDigit[-1]
            hexList.append(hexString + secondDigit[j] + firstDigit)
            if secondDigit[j] == 'a' or secondDigit[j] == 'A':
                num = 160
            elif secondDigit[j] == 'b' or secondDigit[j] == 'B':
                num = 176
            elif secondDigit[j] == 'c' or secondDigit[j] == 'C':
                num = 192
            elif secondDigit[j] == 'd' or secondDigit[j] == 'D':
                num = 208
            elif secondDigit[j] == 'e' or secondDigit[j] == 'E':
                num = 224
            elif secondDigit[j] == 'f' or secondDigit[j] == 'F':
                num = 240
            else:
                num = 16*int(secondDigit[j])
            number = num + i
     
            if secondDigit[j] == '0':
                binString = bin(number)
                while len(binString) != 6:
                    binString = binString[:2] + '0' + binString[2:]
                binaryNum = '0b0000 ' + binString[2:]
            else:
                binaryNum = bin(number)
                binaryNum = binaryNum[:6] + ' ' + binaryNum[6:]
            binList.append(binaryNum)
            aList.append([hexList[j*16 + i], binList[j* 16 + i], number])
            binHexList.append(aList[count])
            count +=1
    #pick twelve pairs for the board
    nextList = []
    for i in range(12):
        randNumber = randint(0, (len(binHexList)-1))
        nextList.append(binHexList.pop(randNumber))

    # now add WildCard
    wildCardNum = randint(0, 23)
    finalTupleList = []
    counter = 0
    done = 0
    for i in range(12):
        if i == int(wildCardNum/2) and done == 0 or i == (int(wildCardNum/2) + 1) and done == 0:
            finalTupleList.append((randomBoxStringTupleList[counter][0], randomBoxStringTupleList[counter][1], 'Wild Card', 1000))
        
            counter +=1
            done = 1
        finalTupleList.append((randomBoxStringTupleList[counter][0], randomBoxStringTupleList[counter][1], nextList[i][0], nextList[i][2]))
        counter +=1
        finalTupleList.append((randomBoxStringTupleList[counter][0], randomBoxStringTupleList[counter][1], nextList[i][1], nextList[i][2]))
        counter +=1
    finalTupleList.sort()
    return finalTupleList


Python/asciiBoard.py
import binHex

sqWid = 13
sqsWide = 5

def repeat(x, n):
    return n*x

rowDiv = [repeat('-', sqWid) for i in range(sqsWide)]
rowDwn = [repeat('X', sqWid) for i in range(sqsWide)]
rowUp = [repeat(' ', sqWid) for i in range(sqsWide)]

def labelDown(label):
    return '{:X^13}'.format(' ' + label + ' ')

def labelUp(label):
    return '{:^13}'.format(' ' + label + ' ')

def row(labels, div, w, n):
    return div+div.join(labels)+div

r0 = [x[2] for x in binHex.binHexNums()[:5]]
r1 = [x[2] for x in binHex.binHexNums()[5:10]]
r2 = [x[2] for x in binHex.binHexNums()[10:15]]
r3 = [x[2] for x in binHex.binHexNums()[15:20]]
r4 = [x[2] for x in binHex.binHexNums()[20:25]]

def printRow(r):
    print(row(rowDiv, "+", 13, 5))
    print(row(rowUp, "|", 13, 5))
    print(row(rowUp, "|", 13, 5))
    print(row(list(map(labelUp, r)), '|', sqWid, sqsWide))
    print(row(rowUp, "|", 13, 5))
    print(row(rowUp, "|", 13, 5))

def printBoard():
    printRow(r0)
    printRow(r1)
    printRow(r2)
    printRow(r3)
    printRow(r4)
    print(row(rowDiv, "+", 13, 5))

printBoard()

Python/startingData.py
from random import randint

def startingData():
    #print()
    boxStringTupleList = []
    nextList = []
    number = 0
    for l in range(5):
        for m in range(5):
            boxStringTupleList.append((number, str(l) + str(m)))
            number +=1
    if (len(boxStringTupleList) != 25):
        print('length of boxStringTupleList = {}'.format(len(boxStringTupleList)))
              
    #print(boxStringTupleList)
    #return finalList

    # make the square numbers random so we can add the random bin and hex values


    #print()
    count = 0
    randomBoxStringTupleList = []
    for i in range(0, (len(boxStringTupleList))):
        number = randint(0, (len(boxStringTupleList)-1))
        randomBoxStringTupleList.append(boxStringTupleList.pop(number))
        count +=1
    if len(randomBoxStringTupleList) != 25:
        print('length of randomBoxStringTupleList = {}'.format(len(randomBoxStringTupleList)))


    ##print('randomBoxStringTupleList = {}'.format(randomBoxStringTupleList))
    ##print('length of RBSTL is {}'.format(len(randomBoxStringTupleList)))

        
    #def listCreator():
        # create possible entries for square
    hexList = []
    binList = []
    binHexList = []
    aList = []
    hexString = '0x'
    secondDigit = ['0', 'a', 'f']
    num = 0
    count = 0

    for j in range(len(secondDigit)):
        for i in range(16):
            firstDigit = str(hex(i))
            firstDigit = firstDigit[-1]
            hexList.append(hexString + secondDigit[j] + firstDigit)
            if secondDigit[j] == 'a' or secondDigit[j] == 'A':
                num = 160
            elif secondDigit[j] == 'b' or secondDigit[j] == 'B':
                num = 176
            elif secondDigit[j] == 'c' or secondDigit[j] == 'C':
                num = 192
            elif secondDigit[j] == 'd' or secondDigit[j] == 'D':
                num = 208
            elif secondDigit[j] == 'e' or secondDigit[j] == 'E':
                num = 224
            elif secondDigit[j] == 'f' or secondDigit[j] == 'F':
                num = 240
            else:
                num = 16*int(secondDigit[j])
            number = num + i
            #print('binNumber = {}, secondDigit = {}'.format(bin(number), secondDigit[j]))
            if secondDigit[j] == '0':
                binString = bin(number)
                while len(binString) != 6:
                    binString = binString[:2] + '0' + binString[2:]
                binaryNum = '0b0000 ' + binString[2:]
            else:
                binaryNum = bin(number)
                binaryNum = binaryNum[:6] + ' ' + binaryNum[6:]
            binList.append(binaryNum)
            aList.append([hexList[j*16 + i], binList[j* 16 + i], number])
            binHexList.append(aList[count])
            count +=1
    #print(binHexList) # this is an ordered sorted list

    #pick twelve pairs for the board
    nextList = []
    for i in range(12):
        randNumber = randint(0, (len(binHexList)-1))
        nextList.append(binHexList.pop(randNumber))
    ##print('nextList = {}'.format(nextList)) # nextList has 12 bin hex pairs
    ##print('length of nextList = {}'.format(len(nextList)))

    # now add WildCard
    wildCardNum = randint(0, 23)
    finalTupleList = []
    counter = 0
    done = 0
    for i in range(12):
        if i == int(wildCardNum/2) and done == 0 or i == (int(wildCardNum/2) + 1) and done == 0:
            finalTupleList.append((randomBoxStringTupleList[counter][0], randomBoxStringTupleList[counter][1], 'Wild\nCard', 1000))
            ##        print('wildCardNum = {}'.format(wildCardNum))
            ##        print('wild = {}'.format(finalTupleList[counter]))
            counter +=1
            done = 1
        finalTupleList.append((randomBoxStringTupleList[counter][0], randomBoxStringTupleList[counter][1], nextList[i][0], nextList[i][2]))
        counter +=1
        finalTupleList.append((randomBoxStringTupleList[counter][0], randomBoxStringTupleList[counter][1], nextList[i][1], nextList[i][2]))
        counter +=1
    #    print('i = {}'.format(i))
    ##    print('counter = {}'.format(counter))
    ##    print('length of finalTupleList = {}'.format(len(finalTupleList)))
    finalTupleList.sort()
    #print(finalTupleList)
    return finalTupleList

'''

for f in range(100):
    tupleList,  wildCardNum = startingData()
    if (len(tupleList) != 25):
        print('length of startingData = {} wildCardNum = {}'.format(len(tupleList), wildCardNum))
        print('list = {}'.format(tupleList))
    
    print('wildCardNum = {}'.format(wildCardNum), end='')
'''


Python/drawBoardLarger.py
import turtle
from turtle import Screen
from turtle import *

class drawBoard():
    screen = Screen()
    screen.setup(width=800, height=800, startx=10, starty=10)


    # Draw large off-white square
    turtle.hideturtle()
    turtle.speed(0)
    turtle.penup()
    turtle.goto(-350, -350)
    color('#000000', '#e4e4ff')
    begin_fill()
    turtle.pendown()
    forward(700)
    left(90)
    forward(700)
    left(90)
    forward(700)
    left(90)
    forward(700)
    end_fill()
    turtle.penup()


    #draw vertical lines
    turtle.goto(-210, 350)
    turtle.pendown()
    turtle.goto(-210, -350)
    turtle.penup()

    turtle.goto(-70, 350)
    turtle.pendown()
    turtle.goto(-70, -350)
    turtle.penup()

    turtle.goto(70, 350)
    turtle.pendown()
    turtle.goto(70, -350)
    turtle.penup()

    turtle.goto(210, 350)
    turtle.pendown()
    turtle.goto(210, -350)
    turtle.penup()

    # draw horizonal lines
    turtle.goto(-350, -210)
    turtle.pendown()
    turtle.goto(350, -210)
    turtle.penup()

    turtle.goto(-350, -70)
    turtle.pendown()
    turtle.goto(350, -70)
    turtle.penup()

    turtle.goto(-350, 70)
    turtle.pendown()
    turtle.goto(350, 70)
    turtle.penup()

    turtle.goto(-350, 210)
    turtle.pendown()
    turtle.goto(350, 210)
    turtle.penup()
    #print('Board Drawn')



concentrationFinal.py
'''
still erases some one or two of the solved values 2-25-2020

'''

from random import randint
import turtle
from turtle import *
import tkinter.font
from tkinter import *
import sys

import drawBoardLarger  # draws game board
import startingData   # gets 12 binary and matching hexadecimal numbers


screen = Screen()
screen.setup(width=1200, height=800, startx=10, starty=10)
index = []
eraseble = []
t = turtle.Turtle()
t.hideturtle()
turtle.speed(0)
drawBoardLarger
CList = startingData.startingData()
count1 = 0
for i in CList:
        indexEntry = [CList[count1][0], CList[count1][1]]
        index.append(indexEntry)   # index[0][0] = 0, type(index[0][0] = <class 'int'>)
        count1 +=1

matchCount = 0
pastGuesses = 0
listofPossibles = ['00', '01', '02', '03', '04', '10', '11', '12', '13', '14', '20', '21', '22', '23', '24', '30', '31', '32', '33', '34', '40', '41', '42', '43', '44']
thisGameListofPs = listofPossibles[:]


def erasableWrite(tortoise, name, font, align, reuse=None):
    # tracks place numbers and values so they can be removed and replaced
    eraser = turtle.Turtle() if reuse is None else reuse
    eraser.hideturtle()
    eraser.up()
    eraser.setposition(tortoise.position())
    eraser.write(name, font=font, align=align)
    return eraser

def printBoard():
    #  Puts place number in the board
    global t, eraseble
    counter = 0
    startX = -280
    startY = -295
    textValue = ''
    for i in range(5):
        for k in range(5):
            textValue = CList[counter][1]
            t.penup()
            t.goto(startX, startY)
            t.pendown()
            eraseble.append(erasableWrite(t, textValue, font=("New Times Roman", 18, "normal"), align="center"))
            startX += 140
            counter +=1
        startY += 140
        startX = -280

def getFirst():
    # gets the place number of the first guess to reveal the value
    textInput = screen.textinput("Concentration Game", "type your first choice")
    return textInput

def getSecond():
    # gets the place number for the second guess to reveal the value
    textInput = screen.textinput("Concentration Game", "type your second choice")
    return textInput

def validate(stringNumber):
    # checks to see if the place number is a valid guess
    global thisGameListofPs
    if stringNumber in thisGameListofPs:
        return stringNumber, False
    else:
        return "", True 


printBoard()  # calls the function and displays the place numbers

def flipCard(choice):
    #  takes the place number and replaces it with the value (0b or 0x)
    global gameTupleList, index, eraseble, CList
    gameTupleList = CList[:]
    firstInt = 0
    counti = 0
    count1 = 0
    textValue = ''
    for x in index:
        if x[1] == choice:
            firstInt = index[counti][0]
        counti +=1
    newEntry = (gameTupleList[firstInt])
    nextStep = (newEntry[0], newEntry[2], newEntry[1], newEntry[3])
    gameTupleList[firstInt] = nextStep
    counter = newEntry[0]
    counterX = counter % 5
    counterY = int(counter/5)
    startX = -280
    startY = -295
    startY += 140 * counterY
    startX += 140 * counterX
    t.penup()
    t.goto(startX, startY)
    print('eraseble 1 at counter {}'.format(counter))
    eraseble[counter].clear()
    textValue = nextStep[1]
    if len(textValue) <= 3:
        erasableWrite(t, textValue, font=("New Times Roman", 18, "normal"), align="center", reuse= eraseble[counter])
    else:
        erasableWrite(t, textValue, font=("New Times Roman", 14, "normal"), align="center", reuse= eraseble[counter])
    return nextStep

    
def gotAMatch():
    messagebox.showinfo('A Match!', 'Excellent : > )')
    
def notAMatch():
    messagebox.showinfo('NOT a match!', 'sorry : > (')

def nextLoop():
    global choice1, choice2, matchCount, CList, thisGameListofPs, gameTupleList, index, eraseble, pastGuesses 
    while matchCount <= 11:
        notValid = True
        while notValid:
            choice1 = getFirst()
            choice1, notValid = validate(choice1)
        card1 = flipCard(choice1)
        notValid = True
        while notValid:
            choice2 = getSecond()
            choice2, notValid = validate(choice2)
        card2 = flipCard(choice2)
        pastGuesses +=1
        if (card1[3] == 1000) or (card2[3] == 1000):
            countLOP = 0
            for n in thisGameListofPs:
                if n == choice1:
                    del thisGameListofPs[countLOP]
                countLOP +=1
            countLOP = 0
            for n in thisGameListofPs:
                if n == choice2:
                    del thisGameListofPs[countLOP]
                countLOP +=1
            if card1[3] == 1000:
                one2M = card2[3]
                cardDone = card2[0]
                for n in CList:
                    if n[3] == card2[3] and n[0] != card2[0]:
                            removePossible = n[1]
            else:
                one2M = card1[3]
                cardDone = card2[0]
                for n in CList:
                    if n[3] == card1[3] and n[0] != card1[0]:
                            removePossible = n[1]
            counter = 0
            for n in CList:
                if  one2M == n[3] and n[0] != cardDone:
                    print('eraseble 2 at n[0] {}'.format(n[0]))
                    eraseble[n[0]].clear()
                    textValue = n[2]
                    counterX = n[0] % 5
                    counterY = int(n[0]/5)
                    startX = -280
                    startY = -295
                    startY += 140 * counterY
                    startX += 140 * counterX
                    t.penup()
                    t.goto(startX, startY)
                    if len(textValue) <= 3:
                        erasableWrite(t, textValue, font=("New Times Roman", 18, "norl"), align="center", reuse= eraseble[counter])
                    else:
                        erasableWrite(t, textValue, font=("New Times Roman", 14, "normal"), align="center", reuse= eraseble[counter])
                    countLOP = 0
                    for m in thisGameListofPs:
                        if m == removePossible:
                            del thisGameListofPs[countLOP]
                            countLOP +=1
                        else:
                            countLOP +=1
                    counter +=1
            gotAMatch()
            matchCount += 1
            CList = gameTupleList[:]
        elif (card1[3] == card2[3]):
            countLOP = 0
            for n in thisGameListofPs:
                if n == choice1:
                    del thisGameListofPs[countLOP]
                countLOP +=1
            countLOP = 0
            for n in thisGameListofPs:
                if n == choice2:
                    del thisGameListofPs[countLOP]
                countLOP +=1
            gotAMatch()
            matchCount += 1
            CList = gameTupleList[:]
        else:
            notAMatch()
            listReset = [choice1, choice2]
            counter = 0
            for a in listReset:
                gameTupleList = CList[:]
                for item in index:
                    if item[1] == a:
                        counter = item[0]
                counterX = counter % 5
                counterY = int(counter/5)
                startX = -280
                startY = -295
                startY += 140 * counterY
                startX += 140 * counterX
                t.penup()
                t.goto(startX, startY)
                print('eraseble 3 at counter {}'.format(counter))
                eraseble[counter].clear()
                textValue = gameTupleList[counter][1]
                erasableWrite(t, textValue, font=("New Times Roman", 18, "normal"), align="center", reuse=eraseble[counter])

    messageFinal = 'In only ' + str(pastGuesses) +' guesses!'
    messagebox.showinfo('Congratulations! You solved the game!', messageFinal)
    continueGame = screen.textinput("Play Again? Yes or No", "Type y or n")
    if continueGame.lower() == 'n':
        screen.bye()
        sys.exit()
        
    elif continueGame.lower() == 'y':
        turtle.clearscreen()
        CList = startingData.startingData()
        thisGameListofPs = listofPossibles[:]
        drawBoardLarger
        eraseble = []
        printBoard()
        pastGuesses = 0
        matchCount = 0
        nextLoop()

nextLoop()

mainloop()
